File: frontend/src/api/pipeline.ts
import request from './request'

export interface StrategyTypeMeta {
  type_key: string
  display_name: string
  description: string
  param_schema: Record<string, unknown>
}

export interface PipelineStep {
  id: number
  strategy_type: string
  display_name: string
  params: Record<string, unknown>
  sort_order: number
  enabled: boolean
}

export const pipelineApi = {
  strategyTypes: () => request.get<unknown, StrategyTypeMeta[]>('/pipeline/strategy-types'),
  listSteps: () => request.get<unknown, PipelineStep[]>('/pipeline/steps'),
  createStep: (body: Omit<PipelineStep, 'id'>) =>
    request.post<unknown, PipelineStep>('/pipeline/steps', body),
  patchStep: (id: number, body: Partial<PipelineStep>) =>
    request.patch<unknown, PipelineStep>(`/pipeline/steps/${id}`, body),
  deleteStep: (id: number) =>
    request.delete<unknown, { deleted: number }>(`/pipeline/steps/${id}`),
  reorder: (order: Array<{ id: number; sort_order: number }>) =>
    request.post<unknown, { reordered: number }>('/pipeline/steps/reorder', { order }),
}

File: backend/app/schemas/pipeline.py
from typing import Any

from pydantic import BaseModel, Field


class StrategyTypeMeta(BaseModel):
    type_key: str
    display_name: str
    description: str
    param_schema: dict


class PipelineStepIn(BaseModel):
    strategy_type: str
    display_name: str = Field(min_length=1, max_length=128)
    params: dict[str, Any] = Field(default_factory=dict)
    sort_order: int = 0
    enabled: bool = True


class PipelineStepPatchIn(BaseModel):
    display_name: str | None = Field(default=None, max_length=128)
    params: dict[str, Any] | None = None
    sort_order: int | None = None
    enabled: bool | None = None


class PipelineStepOut(BaseModel):
    id: int
    strategy_type: str
    display_name: str
    params: dict
    sort_order: int
    enabled: bool

    model_config = {"from_attributes": True}


class ReorderIn(BaseModel):
    order: list[dict]   # [{id, sort_order}]
